Sources/StreamChat/Query/Filter+Predicate.swift

```diff
@@ -0,0 +1,88 @@
+            return NSPredicate(format: "\(key) \(compareToken) %@", valueToUse)
+        }
+
+        let makeCompoundPredicate: (_ logicalType: NSCompoundPredicate.LogicalType) -> NSPredicate? = { logicalType in
+            guard let value = self.value as? [Filter] else { return unableToResolve() }
+            let predicates = value.compactMap(\.predicate)
+            guard !predicates.isEmpty else { return unableToResolve() }
+            return NSCompoundPredicate(type: logicalType, subpredicates: predicates)
+        }
+
+        switch op {
+        case .equal:
+            return makeComparablePredicate("==")
+        case .notEqual:
+            return makeComparablePredicate("!=")
+        case .greater:
+            return makeComparablePredicate(">")
+        case .greaterOrEqual:
+            return makeComparablePredicate(">=")
+        case .less:
+            return makeComparablePredicate("<")
+        case .lessOrEqual:
+            return makeComparablePredicate("<=")
+        case .in:
+            guard let key = dtoKey else { return unableToResolve() }
+            guard let valueArray = value as? [FilterValue] else { return unableToResolve() }
+            return NSCompoundPredicate(
+                type: .and,
+                subpredicates: valueArray.map {
+                    NSPredicate(format: "ANY \(key) == %@", $0.description)
+                }
+            )
+        case .notIn:
+            guard let key = dtoKey else { return unableToResolve() }
+            guard let valueArray = value as? NSArray else { return unableToResolve() }
+            return NSCompoundPredicate(format: "NONE \(key) IN %@", valueArray)
+        case .contains:
+            return makeComparablePredicate("IN")
+        case .exists:
+            guard let key = dtoKey else { return unableToResolve() }
+            guard let booleanValue = value as? Bool else { return unableToResolve() }
+            let compareToken = booleanValue == true ? "!=" : "=="
+            return NSPredicate(format: "\(key) \(compareToken) nil")
+        case .and:
+            return makeCompoundPredicate(.and)
+        case .or:
+            return makeCompoundPredicate(.or)
+        case .nor:
+            guard let value = self.value as? [Filter] else { return unableToResolve() }
+            let predicates = value.compactMap {
+                $0.predicate.map(NSCompoundPredicate.init(notPredicateWithSubpredicate:))
+            }
+            guard !predicates.isEmpty else { return unableToResolve() }
+            return NSCompoundPredicate(type: .and, subpredicates: predicates)
+        case .query:
+            guard let key = dtoKey else { return unableToResolve() }
+            let valueToUse: CVarArg = autoKey?.value ?? value.description
+            return NSPredicate(format: "\(key) CONTAINS %@", valueToUse)
+        case .autocomplete:
+            return unableToResolve()
+        }
```

Comment:
After talking with the backend team, the `.query` I don't think we should support it, we should just return `nil` in this case. This is basically used for search, and in this case, the automatic filter has no use. Besides that, the `contains` is a pretty naive implementation which does not replicate what the server is doing. 

On the other hand, `autocomplete` we can actually support it, and this one is useful. What we should use is the `PREFIX` statement in the NSPredicate. 👍